service/history/ndc_task_util.go

```diff
@@ -235,33 +247,52 @@ func validateTaskGeneration(mutableState workflow.MutableState, taskID int64) er
 	return nil
 }
 
-func getTransferTaskEventIDAndRetryable(
+func transferTaskMutableStateStaleChecker(
 	transferTask tasks.Task,
 	executionInfo *persistencespb.WorkflowExecutionInfo,
-) (int64, bool) {
-	eventID := tasks.GetTransferTaskEventID(transferTask)
-	retryable := true
+) bool {
+
+	// Check to see if mutable state cache needs to be reloaded from a database.
+	// The exception is a transient workflow task that doesn't generate events.
 
-	if task, ok := transferTask.(*tasks.WorkflowTask); ok {
-		retryable = !(executionInfo.WorkflowTaskScheduledEventId == task.ScheduledEventID && executionInfo.WorkflowTaskAttempt > 1)
+	wt, isWt := transferTask.(*tasks.WorkflowTask)
+	if !isWt {
+		return true
 	}
 
-	return eventID, retryable
+	transientWorkflowTask := executionInfo.WorkflowTaskAttempt > 1
+	if transientWorkflowTask && executionInfo.WorkflowTaskScheduledEventId == wt.ScheduledEventID {
+		return false
+	}
+
+	return true
 }
 
-func getTimerTaskEventIDAndRetryable(
+func timerTaskMutableStateStaleChecker(
 	timerTask tasks.Task,
 	executionInfo *persistencespb.WorkflowExecutionInfo,
-) (int64, bool) {
-	eventID := tasks.GetTimerTaskEventID(timerTask)
-	retryable := true
+) bool {
+
+	// Check to see if mutable state cache needs to be reloaded from a database.
+	// Exceptions are:
+	// 1. Transient workflow task that doesn't generate events.
+	// 2. Speculative workflow task that doesn't generate events.
+
+	wttt, isWttt := timerTask.(*tasks.WorkflowTaskTimeoutTask)
+	if !isWttt {
+		return true
+	}
+
+	if wttt.GetCategory() == tasks.CategoryMemoryTimer {
+		return false
+	}
```

Comment:
This is main fix. `executionInfo` can't be used to determinate if WFT is speculative or not because it can be lost.